tensorflow/python/ops/image_ops_impl.py

```diff
@@ -1541,7 +1542,7 @@ def _gif():
       channels_msg = 'Channels must be in (None, 0, 3) when decoding GIF images'
       assert_channels = control_flow_ops.Assert(good_channels, [channels_msg])
       with ops.control_dependencies([assert_channels]):
-        return gen_image_ops.decode_gif(contents)
+        return math_ops.cast(gen_image_ops.decode_gif(contents), dtype)
```

Comment:
Instead of always inserting the `cast` operation into the graph, let's do it conditionally. One way to achieve this might be to do something like:

```python
def _cast(t):
  if dtype == tf.uint8: return t
  return math_ops.cast(t, dtype)
```

And then use `_cast(gen_image_ops.decode_git(...))` etc.